File: apps/mobile/ios/ClipTapKeyboard/Mappers/CategoryMapper.swift
//
//  CategoryMapper.swift
//  ClipTapKeyboard
//
//  カテゴリMapper（TypeScript版 CategoryMapper.ts と同等）
//  読み取り専用操作のみ実装
//

import Foundation
import SQLite3

class CategoryMapper: BaseMapper {

    // MARK: - Singleton

    static let shared = CategoryMapper()

    // MARK: - Initialization

    private init() {
        super.init(tableName: "categories")
    }

    // MARK: - Read Operations

    /// ID指定でカテゴリを取得
    func getById(_ id: String) -> Category? {
        let query = """
            SELECT id, name, color, sortOrder, createdAt
            FROM \(tableName)
            WHERE id = ?
        """

        return executeQuerySingle(query, parameters: [id]) { statement in
            return self.mapCategory(from: statement)
        }
    }

    /// 全カテゴリを取得
    func getAll() -> [Category] {
        let query = """
            SELECT id, name, color, sortOrder, createdAt
            FROM \(tableName)
            ORDER BY sortOrder ASC
        """

        return executeQuery(query) { statement in
            return self.mapCategory(from: statement)
        }
    }

    /// 名前でカテゴリを取得
    func getByName(_ name: String) -> Category? {
        let query = """
            SELECT id, name, color, sortOrder, createdAt
            FROM \(tableName)
            WHERE name = ?
        """

        return executeQuerySingle(query, parameters: [name]) { statement in
            return self.mapCategory(from: statement)
        }
    }

    // MARK: - Mapping

    /// SQLite結果からCategoryモデルにマッピング
    private func mapCategory(from statement: OpaquePointer) -> Category {
        let id = getString(statement, at: 0) ?? ""
        let name = getString(statement, at: 1) ?? ""
        let color = getString(statement, at: 2)
        let sortOrder = getInt(statement, at: 3)
        let createdAt = getString(statement, at: 4) ?? ""

        return Category(
            id: id,
            name: name,
            color: color,
            sortOrder: sortOrder,
            createdAt: createdAt
        )
    }
}

File: apps/mobile/android/app/src/main/java/com/sikakou/cliptap/mappers/CategoryMapper.kt
package com.sikakou.cliptap.mappers

import android.content.Context
import com.sikakou.cliptap.models.Category
import android.util.Log

/**
 * カテゴリのMapper
 * iOS版のCategoryMapper.swiftと同等の機能を提供
 */
class CategoryMapper private constructor(context: Context) : BaseMapper(context) {

    companion object {
        private const val TAG = "CategoryMapper"

        @Volatile
        private var INSTANCE: CategoryMapper? = null

        fun getInstance(context: Context): CategoryMapper {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: CategoryMapper(context.applicationContext).also {
                    INSTANCE = it
                }
            }
        }
    }

    /**
     * 全カテゴリを取得
     */
    fun getAll(): List<Category> {
        val categories = mutableListOf<Category>()

        val query = """
            SELECT id, name, color, sortOrder, createdAt
            FROM categories
            ORDER BY sortOrder ASC
        """

        val cursor = executeQuery(query)
        cursor.use {
            while (it.moveToNext()) {
                categories.add(
                    Category(
                        id = it.getString(0),
                        name = it.getString(1),
                        color = it.getStringOrNull(2),
                        sortOrder = it.getInt(3),
                        createdAt = it.getString(4)
                    )
                )
            }
        }

        Log.d(TAG, "Loaded ${categories.size} categories")
        return categories
    }

    /**
     * IDでカテゴリを取得
     */
    fun getById(id: String): Category? {
        val query = """
            SELECT id, name, color, sortOrder, createdAt
            FROM categories
            WHERE id = ?
        """

        val cursor = executeQuery(query, arrayOf(id))
        cursor.use {
            if (it.moveToFirst()) {
                return Category(
                    id = it.getString(0),
                    name = it.getString(1),
                    color = it.getStringOrNull(2),
                    sortOrder = it.getInt(3),
                    createdAt = it.getString(4)
                )
            }
        }

        return null
    }
}
